streams/src/main/java/org/apache/kafka/streams/state/internals/ReadonlyPartiallyDeserializedSegmentValue.java

```diff
@@ -0,0 +1,206 @@
+                }
+            }
+            // prep for next iteration
+            currIndex = isAscending ? currIndex - 1 : currIndex + 1;
+        }
+        // search in segment expected to find result but did not
+        return null;
+    }
+
+    private long updateCurrNextTimestamp(final int currIndex, final boolean isAscending) {
+        if (isAscending) {
+            return currIndex == recordNumber - 1 ? nextTimestamp : unpackedTimestampAndValueSizes.get(currIndex + 1).timestamp;
+        } else {
+            return currIndex == 0 ? nextTimestamp : unpackedTimestampAndValueSizes.get(currIndex - 1).timestamp;
+        }
+    }
+
+    private int initializeCumvalueSize(final int index, final int currIndex, final boolean isAscending) {
+        return (index == Integer.MAX_VALUE || (!isAscending && index == 0)) ? 0
+                                                                            : isAscending ? cumulativeValueSizes.get(currIndex + 1)
+                                                                                          : cumulativeValueSizes.get(currIndex - 1);
+    }
+
+    private int initializeCurrentIndex(final int index, final boolean isAscending) {
+        return isAscending && index == Integer.MAX_VALUE ? recordNumber - 1 : index;
+    }
+
+
+    private boolean hasStillRecord(final long currTimestamp, final long currNextTimestamp, final ResultOrder order) {
+        return order.equals(ResultOrder.ASCENDING) ? currNextTimestamp != nextTimestamp : currTimestamp != minTimestamp;
+    }
+
+    private boolean hasBeenDeserialized(final boolean isAscending, final int currIndex) {
+        if (!isAscending) {
+            return currIndex <= deserIndex;
+        }
+        return currIndex >= deserIndex;
+    }
+
+    private int getValueSegmentIndex(final ResultOrder order, final int currentCumValueSize, final int currValueSize) {
+        return order.equals(ResultOrder.ASCENDING) ? valuesStartingIndex + (currentCumValueSize - currValueSize)
+                                                   : segmentValue.length - currentCumValueSize;
+    }
+
+    private int getTimestampIndex(final ResultOrder order, final int currIndex) {
+        return order.equals(ResultOrder.ASCENDING) ? valuesStartingIndex - ((recordNumber - currIndex) * (TIMESTAMP_SIZE + VALUE_SIZE))
+                                                   : 2 * TIMESTAMP_SIZE + currIndex * (TIMESTAMP_SIZE + VALUE_SIZE);
+    }
+
+    private void findValuesStartingIndex() {
+        long currTimestamp = -1;
+        int currIndex = 0;
+        int timestampSegmentIndex = 0;
+        while (currTimestamp != minTimestamp) {
+            timestampSegmentIndex = 2 * TIMESTAMP_SIZE + currIndex * (TIMESTAMP_SIZE + VALUE_SIZE);
+            currTimestamp = ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex);
+            unpackedTimestampAndValueSizes.put(currIndex, new TimestampAndValueSize(currTimestamp, Integer.MIN_VALUE));
```

Comment:
We usually use `-1` to indicate "unknown" (not `MIN_VALUE`)

Reply:
> We usually use `-1` to indicate "unknown" (not `MIN_VALUE`)

In `RocksDBVersionedStoreSegmentValueFormatter` class, `-1` was used for tombstones. That's why I wanted to have something different. should we use `Optional` instead?